File: scripts/set-maven-reactor-version.rb
require "pathname"
require "rexml/document"

repository_root = Pathname(ARGV.fetch(0)).expand_path
new_version = ARGV.fetch(1)
unless new_version.match?(/\A[0-9A-Za-z][0-9A-Za-z.+_-]*\z/)
  warn "Invalid Maven project version: #{new_version}"
  exit 2
end

root_pom = repository_root.join("pom.xml")
unless root_pom.file?
  warn "Maven reactor root does not exist: #{root_pom}"
  exit 1
end

root_document = REXML::Document.new(root_pom.read)
root_project = root_document.root
old_version = root_project.elements["version"]&.text.to_s.strip
if old_version.empty?
  warn "Maven reactor root has no project version: #{root_pom}"
  exit 1
end
if old_version == new_version
  warn "New version matches the current reactor version: #{new_version}"
  exit 2
end
root_group_id = root_project.elements["groupId"]&.text.to_s.strip

excluded_directories = %w[.git .worktrees graphify-out target]
pom_files = Dir.glob(repository_root.join("**/pom.xml")).map { |path| Pathname(path) }.reject do |pom|
  pom.relative_path_from(repository_root).each_filename.any? { |part| excluded_directories.include?(part) || part == "test" }
end.sort
edits = {}
updated_references = 0

pom_files.each do |pom|
  document = REXML::Document.new(pom.read)
  project = document.root
  targets = 0
  targets += 1 if project.elements["version"]&.text.to_s.strip == old_version
  parent = project.elements["parent"]
  targets += 1 if parent && parent.elements["groupId"]&.text.to_s.strip == root_group_id && parent.elements["version"]&.text.to_s.strip == old_version
  targets += 1 if project.elements["properties/jfoundry.version"]&.text.to_s.strip == old_version

  content = pom.read
  occurrences = content.scan(/<(?:version|jfoundry\.version)>#{Regexp.escape(old_version)}<\/(?:version|jfoundry\.version)>/).length
  unless occurrences == targets
    next if targets.zero? && occurrences.zero?

    warn "#{pom.relative_path_from(repository_root)}: unclassified #{old_version} occurrence(s); expected #{targets}, found #{occurrences}"
    exit 1
  end
  next if targets.zero?

  edits[pom] = content.gsub(/(<(?:version|jfoundry\.version)>)#{Regexp.escape(old_version)}(<\/(?:version|jfoundry\.version)>)/, "\\1#{new_version}\\2")
  updated_references += targets
end

if edits.empty?
  warn "No reactor version references matched #{old_version}"
  exit 1
end

edits.each { |pom, content| pom.write(content) }
puts "Updated #{updated_references} reactor version references across #{edits.size} XML POMs: #{old_version} -> #{new_version}"

File: scripts/set-maven-reactor-version.py
#!/usr/bin/env python3

"""Update the version references owned by the Maven reactor."""

from __future__ import annotations

import re
import sys
from pathlib import Path
import xml.etree.ElementTree as ET


VERSION_PATTERN = re.compile(r"<(?:version|jfoundry\.version)>[^<]+</(?:version|jfoundry\.version)>")
VALID_VERSION = re.compile(r"^[0-9A-Za-z][0-9A-Za-z.+_-]*$")
EXCLUDED_DIRECTORIES = {".git", ".worktrees", "graphify-out", "target"}


def local_name(tag: str) -> str:
    return tag.rsplit("}", 1)[-1]


def child(element: ET.Element | None, name: str) -> ET.Element | None:
    if element is None:
        return None
    return next((candidate for candidate in element if local_name(candidate.tag) == name), None)


def text(element: ET.Element | None, name: str) -> str:
    candidate = child(element, name)
    return (candidate.text or "").strip() if candidate is not None else ""


def fail(message: str, code: int = 1) -> None:
    print(message, file=sys.stderr)
    raise SystemExit(code)


def pom_files(root: Path) -> list[Path]:
    return sorted(
        path
        for path in root.rglob("pom.xml")
        if not any(part in EXCLUDED_DIRECTORIES or part == "test" for part in path.relative_to(root).parts)
    )


def main() -> None:
    if len(sys.argv) != 3:
        fail(f"Usage: {Path(sys.argv[0]).name} <repository-root> <new-version>", 2)

    repository_root = Path(sys.argv[1]).expanduser().resolve()
    new_version = sys.argv[2]
    if not VALID_VERSION.fullmatch(new_version):
        fail(f"Invalid Maven project version: {new_version}", 2)

    root_pom = repository_root / "pom.xml"
    if not root_pom.is_file():
        fail(f"Maven reactor root does not exist: {root_pom}")

    try:
        root_project = ET.parse(root_pom).getroot()
    except ET.ParseError as error:
        fail(f"Maven reactor root is not valid XML: {error}")

    old_version = text(root_project, "version")
    if not old_version:
        fail(f"Maven reactor root has no project version: {root_pom}")
    if old_version == new_version:
        fail(f"New version matches the current reactor version: {new_version}", 2)
    root_group_id = text(root_project, "groupId")

    edits: dict[Path, str] = {}
    updated_references = 0
    for pom in pom_files(repository_root):
        content = pom.read_text(encoding="utf-8")
        try:
            project = ET.fromstring(content)
        except ET.ParseError as error:
            fail(f"{pom.relative_to(repository_root)} is not valid XML: {error}")

        targets = 1 if text(project, "version") == old_version else 0
        parent = child(project, "parent")
        if parent is not None and text(parent, "groupId") == root_group_id and text(parent, "version") == old_version:
            targets += 1
        properties = child(project, "properties")
        if text(properties, "jfoundry.version") == old_version:
            targets += 1

        occurrences = sum(1 for match in VERSION_PATTERN.finditer(content) if match.group(0).split(">", 1)[1].split("<", 1)[0] == old_version)
        if occurrences != targets:
            if targets == 0 and occurrences == 0:
                continue
            fail(f"{pom.relative_to(repository_root)}: unclassified {old_version} occurrence(s); expected {targets}, found {occurrences}")
        if targets == 0:
            continue

        old_element = re.compile(
            rf"(<(?:version|jfoundry\.version)>)" + re.escape(old_version) + rf"(</(?:version|jfoundry\.version)>)"
        )
        edits[pom] = old_element.sub(rf"\g<1>{new_version}\g<2>", content)
        updated_references += targets

    if not edits:
        fail(f"No reactor version references matched {old_version}")
    for pom, content in edits.items():
        pom.write_text(content, encoding="utf-8")
    print(f"Updated {updated_references} reactor version references across {len(edits)} XML POMs: {old_version} -> {new_version}")


if __name__ == "__main__":
    main()
